src/test/java/com/amazon/dlic/auth/http/jwt/keybyoidc/SelfRefreshingKeySetTest.java

```diff
@@ -11,65 +11,98 @@
+        Assert.assertThrows(AuthenticatorUnavailableException.class, () -> selfRefreshingKeySet.getKey(null));
 
-        ExecutorService executorService = Executors.newCachedThreadPool();
+    }
 
-        Future<JWK> f1 = executorService.submit(() -> selfRefreshingKeySet.getKey("kid/a"));
+    @Test
+    public void getKey_withInvalidDataShouldReturnBadCredentialException() throws AuthenticatorUnavailableException,
+        BadCredentialsException {
 
-        provider.waitForCalled();
+        Assert.assertThrows(BadCredentialsException.class, () -> selfRefreshingKeySet.getKey("kid/X"));
+    }
 
-        Future<JWK> f2 = executorService.submit(() -> selfRefreshingKeySet.getKey("kid/b"));
+    @Test
+    public void getKeyAfterRefresh_withKidShouldReturnKey() throws AuthenticatorUnavailableException, BadCredentialsException {
+
+        OctetSequenceKey key = (OctetSequenceKey) selfRefreshingKeySet.getKeyAfterRefresh("kid/b");
+        assertThat(TestJwk.OCT_2_K, is(equalTo(key.getKeyValue().decodeToString())));
+    }
 
-        while (selfRefreshingKeySet.getQueuedGetCount() == 0) {
-            Thread.sleep(10);
+    @Test
+    public void getKeyAfterRefresh_queuedGetCountVariableShouldBeZeroWhenFinishWithAllKeyRefreshes() throws InterruptedException,
+        ExecutionException {
+
+        int numThreads = 10;
+        ExecutorService executor = Executors.newFixedThreadPool(numThreads);
+        Object lock = new Object();
+
+        for (int i = 0; i < numThreads; i++) {
+            executor.execute(() -> {
+                synchronized (lock) {
+                    try {
+                        selfRefreshingKeySet.getKeyAfterRefresh("kid/a");
+                    } catch (AuthenticatorUnavailableException e) {} catch (BadCredentialsException e) {}
+                }
+            });
         }
 
-        provider.unblock();
+        executor.shutdown();
+        executor.awaitTermination(1, TimeUnit.SECONDS);
 
-        Assert.assertEquals(TestJwk.OCT_1_K, ((OctetSequenceKey) f1.get()).getKeyValue().decodeToString());
-        Assert.assertEquals(TestJwk.OCT_2_K, ((OctetSequenceKey) f2.get()).getKeyValue().decodeToString());
+        assertThat((int) selfRefreshingKeySet.getRefreshCount(), is(equalTo(numThreads)));
+        assertThat((int) selfRefreshingKeySet.getQueuedGetCount(), is(equalTo((0))));
+    }
 
-        Assert.assertEquals(1, selfRefreshingKeySet.getRefreshCount());
-        Assert.assertEquals(1, selfRefreshingKeySet.getQueuedGetCount());
+    @Test
+    public void getKeyAfterRefresh_withNullKidShouldThrowBadCredentialsException() throws AuthenticatorUnavailableException,
+        BadCredentialsException {
 
+        Assert.assertThrows(BadCredentialsException.class, () -> selfRefreshingKeySet.getKeyAfterRefresh(null));
+    }
+
+    @Test
+    public void getKeyAfterRefresh_withInvalidDataShouldReturnBadCredentialException() throws AuthenticatorUnavailableException,
```

Comment:
`public JWK getKey(String kid) throws AuthenticatorUnavailableException, BadCredentialsException`
and
`public synchronized JWK getKeyAfterRefresh(String kid) throws AuthenticatorUnavailableException, BadCredentialsException`
@DarshitChanpura these methods are different, aren't they?
What I know is that when testing we must test all public methods, so that's what I was trying to do. 
Any guidance here to understand better Test concept?